### PR TITLE
HPCC-14912 Ensure thread hooks are called before query dll is unloaded

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday gavin.halliday@lexisnexis.com

### CODE REVIEW

**Comment by ghalliday** [comment]
@jakesmith I suspect there should be a couple of entries in thor, but I don't know where I would put them.  Please comment (and also review).

File: system/jlib/jthread.hpp [hunk 1]
```diff
@@ -61,6 +61,14 @@ extern jlib_decl unsigned threadLogID();  // for use in logging
 
 typedef void (*ThreadTermFunc)();
 extern jlib_decl ThreadTermFunc addThreadTermFunc(ThreadTermFunc onTerm);
+extern jlib_decl void callThreadTerminationHooks();
+
+//An exception safe way of ensuring that the thread termination hooks are called.
```
Comment (jakesmith) [inline system/jlib/jthread.hpp]: s/exception/exceptionally

File: ecl/eclagent/eclagent.cpp [hunk 2]
```diff
@@ -1864,6 +1864,7 @@ void EclAgent::doProcess()
         {
             MTIME_SECTION(queryActiveTimer(), "Process");
             Owned<IEclProcess> process = loadProcess();
+            QueryTerminationCleanup threadCleanup;
```
Comment (jakesmith) [inline ecl/eclagent/eclagent.cpp]: As far as I understand it, this isn't actually going to used, i.e. there won't be a thread termination hook installed on this thread..
That confused me when reviewing this PR a bit.
In Roxie the equivalent code is on it's own thread

@ghalliday - correct me if I'm wrong.

**Comment by jakesmith** [comment]
@ghalliday - already merged, but looks good

File: ecl/eclagent/eclagent.cpp (hunk 2 above)
Comment (richardkchapman) [inline ecl/eclagent/eclagent.cpp]: The call to process() may install arbitrary hooks, if the ECL uses embedded languages (or unicode).

**Pull request merged**